### ISSUE TITLE
[i18n / l10n] Centralized translation repo for plugins?

### ISSUE THREAD

**Issue by rhansen** [ether/etherpad-lite#4317]
Each plugin that supports l10n has its own language files. For a language translator, it is not practical to maintain translations across all of the plugins because each plugin has its own repository with its own maintainer, and many of the maintainers are not responsive. Registering each plugin with translatewiki.net also seems infeasible.

What do you think about maintaining plugin translations in the etherpad-lite project alongside the translations for etherpad-lite itself? These translations would override the translations provided by the plugin (if any). To get the latest translations for all plugins, a user would only need to update etherpad-lite.

Or perhaps the translations for etherpad-lite plus all plugins could be maintained in a special etherpad-translations project. The translations could be automatically published to npmjs.com whenever there is an update from translatewiki.net.

Of course each plugin would still have to provide l10n IDs and call the underscore function where appropriate. But once that is done, translations could be updated even if a plugin maintainer goes dormant.

Thoughts?

cc @ovari

**Comment by ovari** [comment]
@JohnMcLear @Nikerabbit what are your thought please? Thank you

**Comment by JohnMcLear** [comment]
I'm trying to build tools for Devs that will ease the process of implementing language files.  See bin/plugins

**Comment by JohnMcLear** [comment]
> What do you think about maintaining plugin translations in the etherpad-lite project alongside the translations for etherpad-lite itself? These translations would override the translations provided by the plugin (if any). To get the latest translations for all plugins, a user would only need to update etherpad-lite.

I'm not a fan of this.  I think of plugins like apps, I want them contained but working to good practice.

> Or perhaps the translations for etherpad-lite plus all plugins could be maintained in a special etherpad-translations project. The translations could be automatically published to npmjs.com whenever there is an update from translatewiki.net.

Plugin authors could setup a hook to auto publish on TW commit if they wanted?

**Comment by rhansen** [comment]
> Plugin authors could setup a hook to auto publish on TW commit if they wanted?

It's not about what the plugin authors want, it's about what users of those plugins want. Approximately zero plugin authors care about Hungarian translations, but there are definitely some users that care. Devs tend to not spend time on stuff they don't care about unless (1) a user requests it, and (2) the change is sufficiently simple.  Integrating with translatewiki.net is not trivial. Setting up automated updates is less so. Because of this, I don't see many plugin authors setting up automated translation updates without pressure/leadership from Etherpad devs. Even with good documentation few will bother.

Also: Requiring every plugin author to set up their own TW integration and automation is inefficient. We could just do it once for everyone's benefit.

> I think of plugins like apps, I want them contained but working to good practice.

Does it help to think of an `etherpad-translations` npm package as a plugin for plugins?  :smiley:

**Comment by JohnMcLear** [comment]
The goal is to work with @Nikerabbit to get setting up on translatewiki automated.  Being that I think over 50% of the plugins I'm the author for and that all my plugins have automation once we solve it for my plugins we can build it into the plugin automation logic already existing in Etherpad.

I think it's not gonna be 1.9 but maybe 2.0.  I think I need to write down the logic I expect so people can take on specific tasks.  I will do this asap.

**Comment by rhansen** [comment]
> The goal is to work with @Nikerabbit to get setting up on translatewiki automated.

That would help a lot, as would a GitHub action to automatically bump the version number and publish to npm whenever there is a new commit from translatewiki.net.

> all my plugins have automation

Do you mean automation from translatewiki.net to your repos? Or from your repo to published npm? (Or both?)

**Comment by JohnMcLear** [comment]
I automate a bunch of stuff.  Check out the script 👍

**Comment by ilmartyrk** [comment]
For our project we have developed a solution based on crowdin. I created a plugin https://github.com/citizenos/ep_translations that enables to sync translations from git repo. All translations are in https://github.com/citizenos/etherpad_translations. All translation keys for our plugins and etherpad are added to source.json. This is synced with Crowdin as Crowdin supports github integrations. Now all strings can be translated and reviewed and after they are confirmed Crowdin creates a pull request into github reposistory. If pull is merged into master it is pulled into etherpad instance by ep_translations plugin. Our Crowdin project sits here https://crowdin.com/project/citizen-os-etherpad. We currently have only keys for our own plugins and also some translations override the initial translations in plugins and etherpads main translations. This solution helps us to translate all strings from one place, also we don't need to create a new commit to etherpad or any of the plugins if a string needs a translation or a new language is added. Hope this info helps in developing the centralized solution as it is quite a headache if you need to manage an etherpad instance with a bunch of plugins and multiple languages.

**Comment by JohnMcLear** [comment]
@ilmartyrk That's really cool, I'm +1 sticking w/ translatewiki and @Nikerabbit because they have done a great job so far.

I'm getting the plugins ready now as per https://github.com/ether/etherpad-lite/issues/4337 and the next step will be to give @Nikerabbit some time to come up with a guide/solution to "register" a plugin repo with TW.

**Comment by JohnMcLear** [comment]
bump @Nikerabbit to begin the work here.  The plugins are ready!

**Comment by Nikerabbit** [comment]
@JohnMcLear I'm still struggling to figure out where to get a list of good plugins. You linked some npm page earlier but I don't find it now.

Basically I need a list of repos which are not obsolete (like https://github.com/ether/ep_page_view), have i18n files and have a proper license. Some of those I can easily check with a script, but things like being obsolete or unmaintained not.

I see there is a description in package.json files that I could use. But I don't see a automatic way to get a "pretty name" of the plugin.

**Comment by JohnMcLear** [comment]
@Nikerabbit can you please join our discord channell?  Link on etherpad.org.  will talk it through with you

**Comment by Nikerabbit** [comment]
The invite link https://discord.com/invite/6pnP83 is not working for me.

**Comment by JohnMcLear** [comment]
https://discord.gg/daEjfhw
